master/internal/configpolicy/postgres_task_config_policy_intg_test.go

```diff
@@ -0,0 +1,509 @@
+		Count(ctx)
+	require.NoError(t, err)
+	require.Equal(t, 1, count)
+
+	_, err = db.Bun().NewInsert().Model(ntscTCPs).
+		Where("workspace_id = ?", w.ID).
+		Where("workspace_type = ?", model.NTSCType).
+		Value("invariant_config", "?", string(expInvariantConfig)).
+		Exec(ctx)
+	require.ErrorContains(t, err, "duplicate key value violates unique constraint")
+}
+
+func TestDeleteConfigPolicies(t *testing.T) {
+	ctx := context.Background()
+	require.NoError(t, etc.SetRootPath(db.RootFromDB))
+	pgDB, cleanup := db.MustResolveNewPostgresDatabase(t)
+	defer cleanup()
+	db.MustMigrateTestPostgres(t, pgDB, db.MigrationsFromDB)
+
+	user := db.RequireMockUser(t, pgDB)
+	workspaceIDs := []int32{}
+
+	defer func() {
+		err := db.CleanupMockWorkspace(workspaceIDs)
+		if err != nil {
+			log.Errorf("error when cleaning up mock workspaces")
+		}
+	}()
+
+	tests := []struct {
+		name               string
+		global             bool
+		workloadType       model.WorkloadType
+		hasInvariantConfig bool
+		hasConstraints     bool
+		err                *string
+	}{
+
+		{"ntsc config no constraint", false, model.NTSCType, true, false, nil},
+		{"ntsc config and constraint", false, model.NTSCType, true, true, nil},
+		{"ntsc no config has constraint", false, model.NTSCType, false, true, nil},
+		{
+			"unspecified workload type", false, model.UnknownType, true, true,
+			ptrs.Ptr("invalid workload type"),
+		},
+		{"ntsc no config no constraint", false, model.NTSCType, false, false, nil},
+		{"global ntsc config no constraint", true, model.NTSCType, true, false, nil},
+		{"global ntsc config and constraint", true, model.NTSCType, true, true, nil},
+		{
+			"global unspecified workload type", true, model.UnknownType, true, true,
+			ptrs.Ptr("invalid workload type"),
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			switch test.workloadType {
+			case model.ExperimentType:
+				w, expTCP, _ := CreateMockTaskConfigPolicies(ctx, t, pgDB, user, test.global,
+					test.hasInvariantConfig, test.hasConstraints)
+				if !test.global {
+					workspaceIDs = append(workspaceIDs, int32(w.ID))
+				}
+
+				err := DeleteConfigPolicies(ctx, expTCP.WorkspaceID, test.workloadType)
+				if test.err == nil {
+					require.NoError(t, err)
+				} else {
+					require.ErrorContains(t, err, *test.err)
+				}
+
+			default:
+				w, _, ntscTCP := CreateMockTaskConfigPolicies(ctx, t, pgDB, user, test.global,
+					test.hasInvariantConfig, test.hasConstraints)
+				if !test.global {
+					workspaceIDs = append(workspaceIDs, int32(w.ID))
+				}
+
+				err := DeleteConfigPolicies(ctx, ntscTCP.WorkspaceID, test.workloadType)
+				if test.err == nil {
+					require.NoError(t, err)
+				} else {
+					require.ErrorContains(t, err, *test.err)
+				}
+			}
+		})
+	}
+
+	// Verify that trying to delete task config policies for a nonexistent scope doesn't error out.
+	err := DeleteConfigPolicies(ctx, ptrs.Ptr(-1), model.ExperimentType)
```

Comment:
Why don't we want this function to error out if there is a non-existent scope?

Reply:
Hm! You know, this was my (mistaken) way of getting the test cases to pass instead of adding a check in the function that validates the workspace ID. I'll add this, great catch!

Reply:
Hm, ok I'm actually wondering, is it worth adding an extra select query to check this? Techncially, if the workspace ID is invalid, this is should be found in the API layer before the data gets passed to the db (I guess I think of this layer as solely for performing the requested DB operations and nothing else if that makes sense).
@kkunapuli any opinions here?

Reply:
I think both statements are correct.

I agree that validation on the workspace ID should be done in the API layer. However, wouldn't we also get a postgres error if we attempt to delete something that doesn't exist? (nit: I think here we're more concerned about the scope existing in the config policies table than whether or not the workspace ID is valid.)

Reply:
Agreed! We can't submit a nonexistent workspace ID into the `task_config_policies` table, which is checked in the test cases [here](https://github.com/determined-ai/determined/pull/9915/files#diff-58baddc9d21b606b75e99d38dda67525a00b05be289b98d3b6731cc3bf936f7dR198).
Also, a delete query doesn't fail if nothing gets deleted, which happens when an invalid workspace ID is submitted into the Delete function!